### DISCUSSION TITLE
CIP 40 - Discussions for extension to ODIS for password hashing

### DISCUSSION THREAD

**Discussion by nategraf** [celo-org/celo-proposals#342] [CIP Discussions]
Discussion for https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0040.md

**Comment by YazzyYaz** [comment]
PR of discussion https://github.com/celo-org/celo-proposals/pull/234#pullrequestreview-695541766

**Comment by ankushswar1** [comment]
@nategraf Moving the discussion here per @YazzyYaz's recommendation:

Thanks for clarifying all of my questions 🙏 I think a lot of my feedback/confusion was repeated throughout multiple comments, so I'll try to consolidate some of my lingering thoughts here:

1. I think it would be helpful to clarify upfront that there are a finite set of domain `name`s, each with its own ruleset, and a client cannot arbitrarily specify a new domain `name`. You mention that an unrecognized domain `name` would lead to an error at the HTTP transport layer, but I think it may be helpful to mention this at a high-level in the introduction of what the domain struct does. 

2. I still have a little bit of confusion on this line from the `Key management` section:

> Because ODIS' current OPRF function operates with the same cryptographic primitives as the proposed function, and because its a bad idea in general to reuse keys, implementation of this proposal MUST use a new set of keys, not used in any other service.

I initially interpreted this as each domain requires a new set of keys - I think I may have assumed this because you mentioned that different domains should produce different outputs on the same input, but it wasn't clear whether this was accomplished through a salt or simply different keys - it may be helpful to suggest how that might work as you did [here](https://github.com/celo-org/celo-proposals/pull/234#discussion_r660877917). 

My understanding after reading your responses is that this proposal would only require **one** new DKG, and the same keys will be used for each domain (and thus new domains don't require a new DKG). The new DKG is simply necessary because the proposal modifies ODIS significantly, and because it's a bad idea to reuse keys. Is that correct? It might be helpful to explicitly state that the same keys are used for different domains, and that the distinct outputs of different domains comes from something inherent to the domain struct (e.g. a salt), as opposed to using different signing keys for each domain.

**Comment by nategraf** [comment]
1. Good callout. I've tried to address this in 93a66cca80bdfb3dacb561271bea15a305ca9c03
2. I've opened up a PR to add a note to try to make this a bit more clear. PTAL https://github.com/celo-org/celo-proposals/pull/236
